### ISSUE TITLE
DNS resolution no longer works on iOS with custom DNS provider & Tailscale DNS Settings enabled

### ISSUE THREAD

**Issue by chrigi** [tailscale/tailscale#12563]
### What is the issue?

When connected to Tailscale on iOS with "Use Tailscale DNS Settings" enabled and with a configured custom DNS provider enabled in the iOS settings, iOS won't resolve the Tailscale domain names.

This does sound like "works as intended" but this exact same setup worked until recently. Maybe it relied on some coincidence that on iOS the DNS requests wen't through VPN when it was connected and now don't anymore if a custom DNS is configured?

"Override local DNS" set to "Enabled" is not an option for me as I want to give the devices on the network the freedom to resolve however they want. On the host where the service runs (Windows 10, Tailscale 1.68.1, All preferences checked) I can resolve the Tailscale domain names but all other domains are resolved over the locally configured DNS it seems.

### Steps to reproduce

Setup:
- iOS Settings > VPN & Device Management > Restrictions and Proxies > DNS: NextDNS
- iOS Tailscale App Settings > DNS Settings > Use Tailscale DNS Settings: Enabled
- iOS Tailscale App > Exit Node: none
- Web Tailscale Dashboard > DNS > Nameservers > Override local DNS: Disabled, Additional Global Nameserver configured: NextDNS

1.) Try to open a service through a Tailscale domain (e.g. in Safari)
2.) Notice that it can't resolve the domain
3.) Try using the Tailscale IP of the host on which the service runs
4.) Notice that it works, so the Tailscail connectivity is there
5.) Change any of the following options:
- Exit node enabled
- NextDNS disabled on iOS device (set to "Automatic" in iOS settings)
- "Override local DNS" set to "Enabled" on Tailscale's DNS settings

6.) Notice that setting any of these settings fixes it

### Are there any recent changes that introduced the issue?

This exact setup worked fine for a long time and only broke recently. Until now I could always resolve Tailscale domain names on iOS with these settings.

### OS

iOS

### OS version

17.5.1

### Tailscale version

1.68.0 

### Other software

NextDNS app v2.0.1 (29)

**Comment by agottardo** [comment]
Can you try adding the `disable-split-dns-when-no-custom-resolvers` node attr to your ACLs? It disables a battery life optimization that might be causing this. Then, disconnect and reconnect to Tailscale and see if the issue persists. For instance:

```json
"nodeAttrs": [
		{
			"target": ["autogroup:member"],
			"attr":   ["disable-split-dns-when-no-custom-resolvers"],
		},
]
```

**Comment by agottardo** [comment]
As a quick temporary workaround, I suggest disabling Tailscale DNS and defining custom names in NextDNS for your tailnet devices IPs. That way, all queries will go to NextDNS and Tailscale won't be used.

**Comment by chrigi** [comment]
@agottardo Thanks, the `disable-split-dns-when-no-custom-resolvers` nodeAttr worked. After adding it to my ACL it immediately worked again as previously (with the previously mentioned setup).
What kind of impact does this have? Also on other devices on the network?

**Comment by agottardo** [comment]
The flag should only affect iOS devices.

**Comment by Yeradon** [comment]
I faced a similar issue with my dns setup (global dns (cloudflare) + split dns for my personal domain using NextDNS). Even though the admin interface indicated that this setup wasn't working, it was functioning properly until recently. It seems that activating the ACL noted by @agottardo brings back the original behavior. I am using tailscale 1.70.0 on iOS 17.5.1.

**Comment by dchakro** [comment]
Adding the note attributes as described in https://github.com/tailscale/tailscale/issues/12563#issuecomment-2181565517 worked for me as well. I was unaffected until a couple of weeks back. Out of curiosity, does adding this to my ACL compromises my security or increases my vulnerability?

**Comment by agottardo** [comment]
> Adding the note attributes as described in [#12563 (comment)](https://github.com/tailscale/tailscale/issues/12563#issuecomment-2181565517) worked for me as well. I was unaffected until a couple of weeks back. Out of curiosity, does adding this to my ACL compromises my security or increases my vulnerability?

@dchakro No, adding this to your ACL doesn't compromise your security. It just instructs iOS to route all DNS queries through the Tailscale app, instead of only routing DNS traffic meant for the tailnet. The main impact you might notice is a slight increase in battery usage reported for Tailscale in the Settings app, since Tailscale will now handle all DNS resolution rather than using iOS’s default DNS resolver.

**Comment by dchakro** [comment]
Thank you for the clarification, I will monitor my battery usage to check if I need to find an alternative solution.

**Comment by agottardo** [comment]
@dchakro it shouldn't affect battery _consumption_. It affects the _reported_ battery consumption attributed to the app. Tailscale is handling DNS instead of iOS. Essentially: you're still using the same amount of energy, the system is just attributing it to a different actor :)

**Comment by dchakro** [comment]
Oh That's perfect!

**Comment by larryqiann** [comment]
This worked for me on a device with only Tailscale installed as a VPN. After installing this, HE.net's DNS lookup started to work correctly, and now I can access all of my tailnet services properly again.

Given that my use case is a fairly common one, I think that this should be added to the wiki or even to the app under troubleshooting.

**Comment by FueledByGoogle** [comment]
Any updates on this? iPhone 16 pro max on 26.5 and adding ACL did not fix the issue.
